internal/reconciliations/istio/configuration_test.go

```diff
@@ -1,32 +1,180 @@
 package istio_test
 
 import (
+	"context"
 	"fmt"
+
 	operatorv1alpha1 "github.com/kyma-project/istio/operator/api/v1alpha1"
-	"github.com/kyma-project/istio/operator/internal/reconciliations/istio"
+	istio "github.com/kyma-project/istio/operator/internal/reconciliations/istio"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/types"
+
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/scheme"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 )
 
 const (
 	mockIstioTag             string = "1.16.1-distroless"
 	lastAppliedConfiguration string = "operator.kyma-project.io/lastAppliedConfiguration"
 )
 
-var _ = Describe("CR configuration", func() {
-	Context("UpdateLastAppliedConfiguration", func() {
-		It("should update CR with IstioVersion and spec of CR", func() {
+var _ = Describe("Istio Configuration", func() {
+	Context("LastAppliedConfiguration", func() {
+		It("should update lastAppliedConfiguration and is able to unmarshal it back from annotation", func() {
 			// given
-			cr := operatorv1alpha1.Istio{}
+			numTrustedProxies := 1
+			istioCR := operatorv1alpha1.Istio{Spec: operatorv1alpha1.IstioSpec{Config: operatorv1alpha1.Config{NumTrustedProxies: &numTrustedProxies}}}
 
 			// when
-			updatedCR, err := istio.UpdateLastAppliedConfiguration(cr, mockIstioTag)
+			updatedCR, err := istio.UpdateLastAppliedConfiguration(istioCR, mockIstioTag)
 
 			// then
 			Expect(err).ShouldNot(HaveOccurred())
 			Expect(updatedCR.Annotations).To(Not(BeEmpty()))
+			Expect(updatedCR.Annotations[lastAppliedConfiguration]).To(Equal(fmt.Sprintf(`{"config":{"numTrustedProxies":1},"IstioTag":"%s"}`, mockIstioTag)))
+
+			appliedConfig, err := istio.GetLastAppliedConfiguration(updatedCR)
 
-			Expect(updatedCR.Annotations[lastAppliedConfiguration]).To(Equal(fmt.Sprintf(`{"config":{},"IstioTag":"%s"}`, mockIstioTag)))
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(*appliedConfig.Config.NumTrustedProxies).To(Equal(1))
 		})
 	})
 })
+
+var _ = Describe("Ingress Gateway", func() {
+	Context("IngressGatewayNeedsRestart", func() {
+		It("should restart when CR numTrustedProxies is 2 and in lastAppliedConfig is 1", func() {
+			//given
+			oldNumTrustedProxies := 1
+
+			istioCR := operatorv1alpha1.Istio{}
+			istioCR.Spec.Config.NumTrustedProxies = &oldNumTrustedProxies
+
+			updatedCR, err := istio.UpdateLastAppliedConfiguration(istioCR, mockIstioTag)
```

Comment:
We use the function `UpdateLastAppliedConfiguration` to setup test data without having an explicit test that is testing its working as expected. I think we should have a test verifying the correct behaviour of the function if we want to use it to setup test data. I personally prefer to not rely on other implementation, but create static test data for the test to avoid failing tests, because of bugs in functionality that is not the subject under test.

Reply:
Function is tested in the first test in Context `LastAppliedConfiguration`, but I can agree it is not so nice to reuse other code to setup tests. I just wanted to avoid some code/func duplication. Will redo it.